File: 2022/04/solution.pas
{ solution.pas -- camp cleanup -- Troy Brumley blametroi@gmail.com }

program solution;

{$longstrings off}
{$booleval on}


{ advent of code 2022 day 04 -- camp cleanup

  find overlapping grid assignments and report }


type

   integer = int64;

const

   tGridNumBeg = 0;     { the type includes 'impossible' }
   tGridNumMin = 1;     { values at either end to make   }
   tGridNumMax = 99;    { some code cleaner              }
   tGridNumEnd = 100;   { input validate against min-max }

type

   tGridNum = tGridNumBeg .. tGridNumEnd;
   tGridSet = set of tGridNum;
   tGridAssignment = record
                        spec: string;
                        gridSet: tGridSet;
                     end;

   { a processed line from the input dataset. }

   tPayload = record
                 one,
                 two  : tGridAssignment;
              end;

   { the input dataset for the days problem, one per line of input.
     the original input line and its parsed value(s). blank lines are
     maintained for control break purposes. }

   pData = ^tData;
   tData = record
              fwd,              { a singly linked list would be }
              bwd : pData;      { enough, but you never know }
              num : integer;    { line number in source }
              rec : string;     { original input }
              pld : tPayload;   { what does it mean stimpy? }
           end;


{$I str2gridset.p }
{$I gridset2str.p }
{$I aocinput.p }


{ parse the input line into two gridsets. each line holds two
  assignments to grids specified as ranges: low-high,low-high }

function parse(s : string) : tPayload;

var
   i, j, p : integer;
   go, gt  : tGridAssignment;

begin
   go.spec := '';
   for i := 1 to length(s) do
      if s[i] <> ',' then
         go.spec := go.spec + s[i]
      else
         break;
   p := 1;
   go.gridSet := str2gridset(go.spec, p);
   p := 1;
   gt.spec := '';
   for j := i + 1 to length(s) do
      gt.spec := gt.spec + s[j];
   gt.gridSet := str2gridset(gt.spec, p);
   parse.one := go;
   parse.two := gt;
end;


{ load the dataset into memory. since dynamic arrays aren't standard
  across pascal implementations, and i dislike the implementation in
  free pascal, a doubly linked list is used. it's certainly overkill
  in terms of flexibility, and wouldn't scale well with large
  datasets, but for advent of code a linear search on post 2010
  hardware will be more than fast enough. }

function load : pData;

var
   first,
   prior,
   curr : pData;
   i    : integer;
   s    : string;

begin
   first := nil;
   prior := nil;
   curr := nil;
   i := 0;
   while not eof(input) do begin
      i := i + 1;
      readln(input, s);
      new(curr);
      curr^.fwd := nil;
      curr^.bwd := prior;
      curr^.num := i;
      curr^.rec := s;
      curr^.pld := parse(s);
      if first = nil then
         first := curr
      else
         prior^.fwd := curr;
      prior := curr
   end; { while not eof }

   load := first;
end;


{ see load and parse for more details. count pairs where one range
  completely contains the other. }

function partone(pd : pData) : integer;

begin
   partone := 0;
   while pd <> nil do
      with pd^.pld do begin
         if (one.gridSet <= two.gridSet) or (two.gridSet <= one.gridSet) then
            partone := partone + 1;
         pd := pd^.fwd
      end; { with pd^.pld }
end;


{ see load and parse for more details. count pairs where there is
  any overlap. }

function parttwo(pd : pData) : integer;

begin
   parttwo := 0;
   while pd <> nil do
      with pd^.pld do begin
         if (one.gridSet * two.gridSet) <> [] then
            parttwo := parttwo + 1;
         pd := pd^.fwd
      end; { with pd^.pld }
end;


{ the mainline driver for a day's solution. load and then iterate and
display results. input and output are fixed named files. }

procedure mainline;

var
   dataset  : pData;
   p1answer,
   p2answer : integer;

begin
   aocopen;
   dataset := load;
   p1answer := partone(dataset);
   p2answer := parttwo(dataset);
   writeln('part one answer : ', p1answer);
   writeln('part two answer : ', p2answer);
   aocclose;
end;


{ minimal boot to mainline }

begin
  mainline;
end.

File: 2022/04/solution.f90
! solution.f90 -- 2022 day 04 -- camp cleanup -- troy brumley blametroi@gmail.com

! advent of code 2022 day 04 -- camp cleanup
!
! input is a list of paired integer ranges. the integers have a minimum value of 1
! and a maximum value of 99 and represent 'grid' assignments. in each part of the
! problem we are looking for how each pair may overlap.
!
! for example:
!
! 1-5,4-7  overlap
! 2-8,5-6  one completely contained in the other
! 1-3,7-9  no overlap
!
! for part one, how many assignment pairs does one range fully contain the other?
program solution

   use iso_fortran_env, only: int64
   use aocinput, only: open_aoc_input, read_aoc_input, close_aoc_input, max_aoc_reclen, rewind_aoc_input
   use intset, only: size_of_set, clear_set, include_in_set, exclude_from_set, intersection_of_sets, &
                     union_of_sets, intset_limit, contains_set

   implicit none

   integer, parameter            :: AOCIN = 10
   integer(kind=int64)           :: part_one, part_two  ! results
   integer                       :: i                   ! work
   logical                       :: one(0:intset_limit), two(0:intset_limit), onetwo(0:intset_limit)
   integer                       :: onelo, onehi, twolo, twohi ! range endpoints
   character(len=max_aoc_reclen) :: rec

   ! input file is argument 1
   call open_aoc_input(AOCIN)

   ! initialize
   part_one = 0; part_two = 0

   read_one: do while (read_aoc_input(AOCIN, rec))

      call sanitize(rec)
      read (rec, *) onelo, onehi, twolo, twohi

      call clear_set(one)
      do i = onelo, onehi
         call include_in_set(one, i)
      end do

      call clear_set(two)
      do i = twolo, twohi
         call include_in_set(two, i)
      end do

      if (contains_set(one, two)) then
         part_one = part_one + 1
      else if (contains_set(two, one)) then
         part_one = part_one + 1
      end if

      call intersection_of_sets(one, two, onetwo)

      if (size_of_set(onetwo) /= 0) then
         part_two = part_two + 1
      end if

   end do read_one

   ! report and close
   print *
   print *, part_one, "part one"
   print *, part_two, "part two"
   print *

   call close_aoc_input(AOCIN)

   stop

contains

   ! the input record is a comma delimited list of positive integer ranges.
   ! changing the delimiters to blanks will allow us to use read to get
   ! extract the values.

   subroutine sanitize(ar)
      implicit none
      character(*), intent(inout) :: ar
      integer                     :: vi

      do vi = 1, len(trim(ar))
         if ((ar(vi:vi) == '-') .or. (ar(vi:vi) == ',')) then
            ar(vi:vi) = ' '
         end if
      end do
   end subroutine sanitize

end program solution
!   tGridAssignment = record
!                        spec: string;
!                        gridSet: tGridSet;
!                     end;
!
!
!! see load and parse for more details. count pairs where there is
!! any overlap.
!
!function part_two(pd : pData) : integer;
!
!begin
!   part_two := 0;
!   while pd <> nil do
!      with pd^.pld do begin
!         if (one.gridSet * two.gridSet) <> [] then
!            part_two := part_two + 1;
!         pd := pd^.fwd
!      end; ! with pd^.pld
!end;
